Weird `AssertionError`

```
2025-10-18 18:42:10.294 | INFO     | audiblelight.synthesize:render_audio_for_all_scene_events:571 - Rendered scene audio in 1.89 seconds!
CreateContext: Context created
Generating for test room 14/48, name Kingdom.glb...:  80%|████████████████████████████████████████████████████████████████████████████████▊                    | 4/5 [08:42<02:10, 130.71s/it]
Traceback (most recent call last):
  File "/home/hwc31@cam.ac.uk/AudibleLight/audiblelight/synthesize.py", line 535, in render_audio_for_all_scene_events
    _ = scene.state.irs
        ^^^^^^^^^^^^^^^
  File "/home/hwc31@cam.ac.uk/AudibleLight/audiblelight/worldstate.py", line 457, in irs
    raise AttributeError(
AttributeError: IRs have not been simulated yet: add microphones and emitters and call `simulate`.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/hwc31@cam.ac.uk/AudibleLight/scripts/experiments/dcase_synthetic_data.py", line 329, in <module>
    main(**args)
  File "/home/hwc31@cam.ac.uk/AudibleLight/scripts/experiments/dcase_synthetic_data.py", line 280, in main
    generate(
  File "/home/hwc31@cam.ac.uk/AudibleLight/scripts/experiments/dcase_synthetic_data.py", line 213, in generate
    scene.generate(
  File "/home/hwc31@cam.ac.uk/AudibleLight/audiblelight/core.py", line 1230, in generate
    render_audio_for_all_scene_events(self)
  File "/home/hwc31@cam.ac.uk/AudibleLight/audiblelight/synthesize.py", line 537, in render_audio_for_all_scene_events
    scene.state.simulate()
  File "/home/hwc31@cam.ac.uk/AudibleLight/audiblelight/worldstate.py", line 1739, in simulate
    self._simulation_sanity_check()
  File "/home/hwc31@cam.ac.uk/AudibleLight/audiblelight/worldstate.py", line 1703, in _simulation_sanity_check
    self.num_emitters > 0
AssertionError: Must have added valid emitters to the mesh before calling `.simulate`!
```

--- a/scripts/experiments/generate_dataset.py
+++ b/scripts/experiments/generate_dataset.py
@@ -27,6 +27,7 @@ import random
 from pathlib import Path
 from time import time
 
+import matplotlib.pyplot as plt
 import numpy as np
 from loguru import logger
 from scipy import stats
@@ -49,15 +50,6 @@ OUTPUT_DIR = utils.get_project_root() / "spatial_scenes_dcase_synthetic"
 # Parameters taken from DCASE data
 DURATION = 60
 SAMPLE_RATE = 24000
-MAX_OVERLAP = 2
-
-# Distributions to sample for events
-STATIC_EVENTS = utils.sanitise_distribution(
-    lambda: random.choice(range(config.MIN_STATIC_EVENTS, config.MAX_STATIC_EVENTS))
-)
-MOVING_EVENTS = utils.sanitise_distribution(
-    lambda: random.choice(range(config.MIN_MOVING_EVENTS, config.MAX_MOVING_EVENTS))
-)
 
 # Valid materials for the ray-tracing engine
 with open(MATERIALS_JSON, "r") as js_in:
@@ -67,7 +59,7 @@ VALID_MATERIALS = list({mat["name"] for mat in js_out["materials"]})
 AUGMENTATIONS = {
     "pitchshift": (
         PitchShift,
-        dict(sample_rate=SAMPLE_RATE, semitones=lambda: np.random.randint(-7, 8)),
+        dict(sample_rate=SAMPLE_RATE, semitones=stats.uniform(-7, 0)),
     ),
     "speedup": (
         SpeedUp,
@@ -106,6 +98,11 @@ def generate(
     channel_layout: str,
     augmentations: list[str],
     materials: bool,
+    max_overlap: int,
+    min_events_static,
+    max_events_static: int,
+    min_events_moving: int,
+    max_events_moving: int,
 ) -> None:
     """
     Make a single generation with required arguments
@@ -126,7 +123,7 @@ def generate(
             metadata_path.stem + f"_{channel_layout}.csv"
         ).exists()
     ):
-        return
+        return None
 
     # Choose a material to use
     if materials:
@@ -142,6 +139,14 @@ def generate(
         use_augmentations = get_augmentations(augmentations)
     else:
         use_augmentations = None
+
+    # Distributions to sample for events
+    static_events = utils.sanitise_distribution(
+        lambda: random.choice(range(min_events_static, max_events_static + 1)),
+    )
+    moving_events = utils.sanitise_distribution(
+        lambda: random.choice(range(min_events_moving, max_events_moving + 1)),
+    )
 
     scene = Scene(
         duration=DURATION,
@@ -171,7 +176,7 @@ def generate(
         # Event augmentations will sample from this list
         event_augmentations=use_augmentations,
         fg_path=Path(FG_DIR),
-        max_overlap=MAX_OVERLAP,
+        max_overlap=max_overlap,
         ref_db=scene_ref_db,
         state_kwargs=dict(
             add_to_context=False,
@@ -187,7 +192,7 @@ def generate(
         microphone_type="ambeovr" if channel_layout == "mic" else "foalistener",
         alias=channel_layout,
     )
-    for _ in range(STATIC_EVENTS.rvs()):
+    for _ in range(static_events.rvs()):
         try:
             scene.add_event(
                 event_type="static",
@@ -198,7 +203,7 @@ def generate(
         except ValueError as e:
             logger.warning(e)
 
-    for _ in range(MOVING_EVENTS.rvs()):
+    for _ in range(moving_events.rvs()):
         # Sample the shape to use for this moving event: one of random walk, semicircular, linear
         shape = random.choice(config.MOVING_EVENT_SHAPES)
         try:
@@ -215,9 +220,9 @@ def generate(
     # Always add gaussian noise
     scene.add_ambience(noise="gaussian")
 
-    # If no events added successfully, try again
+    # If no events added successfully, try again by calling the function recursively
     if len(scene.get_events()) == 0:
-        generate(
+        return generate(
             mesh_name,
             split=split,
             scene_num=scene_num,
@@ -226,24 +231,45 @@ def generate(
             channel_layout=channel_layout,
             augmentations=augmentations,
             materials=materials,
+            max_overlap=max_overlap,
+            min_events_static=min_events_static,
+            min_events_moving=min_events_moving,
+            max_events_static=max_events_static,
+            max_events_moving=max_events_moving,
         )
 
     # Do the generation: create audio and DCASE metadata
-    scene.generate(
-        audio_fname=audio_path,
-        metadata_fname=metadata_path,
-        audio=True,
-        metadata_json=True,
-        metadata_dcase=True,
-    )
+    else:
+        scene.generate(
+            audio_fname=audio_path,
+            metadata_fname=metadata_path,
+            audio=True,
+            metadata_json=True,
+            metadata_dcase=True,
+        )
 
-    # Also dump an image of the state
-    fig = scene.state.create_plot()
-    fig.savefig(metadata_path.with_suffix(".png").as_posix())
+        # Also dump an image of the state
+        fig = scene.state.create_plot()
+        fig.savefig(metadata_path.with_suffix(".png").as_posix())
+
+        # Make sure to close the figure!
+        fig.clear()
+        plt.close(fig)
+
+        return None
 
 
 def main(
-    channel_layout: str, augmentations: bool, materials: bool, meshes: int, outdir: str
+    channel_layout: str,
+    augmentations: bool,
+    materials: bool,
+    meshes: str,
+    outdir: str,
+    max_overlap: int,
+    min_events_static,
+    max_events_static: int,
+    min_events_moving: int,
+    max_events_moving: int,
 ):
     # Parse the channel layout and microphone type
     if channel_layout not in ["mic", "foa"]:
@@ -288,6 +314,11 @@ def main(
                 channel_layout,
                 augmentations,
                 materials,
+                max_overlap,
+                min_events_static,
+                max_events_static,
+                min_events_moving,
+                max_events_moving,
             )
 
     logger.info("Generating testing scenes...")
@@ -305,6 +336,11 @@ def main(
                 channel_layout,
                 augmentations,
                 materials,
+                max_overlap,
+                min_events_static,
+                max_events_static,
+                min_events_moving,
+                max_events_moving,
             )
 
     # Log the time taken
@@ -334,9 +370,10 @@ if __name__ == "__main__":
     )
     parser.add_argument(
         "--meshes",
-        type=int,
-        default=9,
-        help="The number of meshes to use. Note that the total number of scapes to generate will remain fixed at 1200.",
+        type=str,
+        default="9A",
+        help="The split of meshes to use: see `dcase_selected_meshes.py`. "
+        "Note that the total number of scapes to generate will remain fixed at 1200.",
     )
     parser.add_argument(
         "--outdir",
@@ -344,6 +381,37 @@ if __name__ == "__main__":
         default=OUTPUT_DIR,
         help=f"Path to save generated outputs, defaults to {OUTPUT_DIR}",
     )
+    parser.add_argument(
+        "--max-overlap",
+        type=int,
+        default=config.MAX_OVERLAP,
+        help=f"Maximum number of overlapping events, defaults to {config.MAX_OVERLAP}",
+    )
+    parser.add_argument(
+        "--min-events-static",
+        type=int,
+        default=config.MIN_STATIC_EVENTS,
+        help=f"Minimum number of static events per scene, defaults to {config.MIN_STATIC_EVENTS}",
+    )
+    parser.add_argument(
+        "--max-events-static",
+        type=int,
+        default=config.MAX_STATIC_EVENTS,
+        help=f"Maximum number of static events per scene, defaults to {config.MAX_STATIC_EVENTS}",
+    )
+    parser.add_argument(
+        "--min-events-moving",
+        type=int,
+        default=config.MIN_MOVING_EVENTS,
+        help=f"Minimum number of moving events per scene, defaults to {config.MIN_MOVING_EVENTS}",
+    )
+    parser.add_argument(
+        "--max-events-moving",
+        type=int,
+        default=config.MAX_MOVING_EVENTS,
+        help=f"Maximum number of static events per scene, defaults to {config.MAX_MOVING_EVENTS}",
+    )
+
     args = vars(parser.parse_args())
     logger.info("Generating with args: {}".format(args))
 

